src/chain_sync/tipset_syncer.rs

```diff
@@ -822,7 +822,7 @@ async fn sync_headers_in_reverse<DB: Blockstore + Sync + Send + 'static>(
 
     #[allow(deprecated)] // Tracking issue: https://github.com/ChainSafe/forest/issues/3157
     let wp = WithProgressRaw::new("Downloading headers", total_size as u64);
-    while pending_tipsets.last().epoch() > until_epoch {
+    'outer: while pending_tipsets.last().epoch() > until_epoch {
```

Comment:
When the `until_epoch` is null, it relies on `break 'outer;` to break this loop

Reply:
Could we avoid the inner/outer loop complexity by refactoring this to smaller functions that could potentially be unit tested? At least the block `for tipset in network_tipsets {`

Reply:
To avoid the `'outer` label, how about this
```rust
// breaks the loop when `until_epoch` has been added to `pending_tipsets`
        // or when `until_epoch` is null. Note that in the latter case, the outer
        // while condition can never be satisfied and it relies on this flag to break
        // the loop
        let should_terminate = 'append_tipsets: {
            for tipset in network_tipsets {
                // This could happen when the `until_epoch` is a null epoch
                if tipset.epoch() < until_epoch {
                    break 'append_tipsets true;
                }
                validate_tipset_against_cache(bad_block_cache, tipset.key(), &accepted_blocks)?;
                accepted_blocks.extend(tipset.cids());
                tracker.write().set_epoch(tipset.epoch());
                pending_tipsets.push(tipset);
            }
            false
        };
        if should_terminate {
            break;
        }
```

Reply:
It's up to you, but I'd still recommend extracting the loop to a separate function. This makes it possible to test and is subjectively easier for the reader of the parent function to grasp.

Reply:
How about this then?
```rust
// tipsets is sorted by epoch in descending order
// returns true when `until_epoch_inclusive` is overreached
fn for_each_tipset_until_epoch(
    tipsets: &[Arc<Tipset>],
    until_epoch_inclusive: ChainEpoch,
    mut callback: impl FnMut(&Arc<Tipset>) -> anyhow::Result<()>,
) -> anyhow::Result<bool> {
    for tipset in tipsets {
        if tipset.epoch() < until_epoch_inclusive {
            return Ok(true);
        }
        callback(tipset)?;
    }
    Ok(false)
}
```

Reply:
LGTM

Reply:
Since this has been queued, I will make this refactor in a separate PR and let's discuss there